app/src/main/java/com/nextcloud/talk/jobs/LeaveConversationWorker.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Nextcloud Talk - Android Client
+ *
+ * SPDX-FileCopyrightText: 2022 Andy Scherzinger <info@andy-scherzinger.de>
+ * SPDX-FileCopyrightText: 2017-2018 Mario Danic <mario@lovelyhq.com>
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+package com.nextcloud.talk.jobs
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.util.Log
+import androidx.work.ListenableWorker
+import androidx.work.WorkerParameters
+import androidx.work.impl.utils.futures.SettableFuture
+import autodagger.AutoInjector
+import com.google.common.util.concurrent.ListenableFuture
+import com.nextcloud.talk.api.NcApi
+import com.nextcloud.talk.application.NextcloudTalkApplication
+import com.nextcloud.talk.models.json.generic.GenericOverall
+import com.nextcloud.talk.users.UserManager
+import com.nextcloud.talk.utils.ApiUtils
+import com.nextcloud.talk.utils.ApiUtils.getConversationApiVersion
+import com.nextcloud.talk.utils.ApiUtils.getCredentials
+import com.nextcloud.talk.utils.ApiUtils.getUrlForParticipantsSelf
+import com.nextcloud.talk.utils.bundle.BundleKeys
+import io.reactivex.Observer
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import javax.inject.Inject
+
+@SuppressLint("RestrictedApi")
+@AutoInjector(NextcloudTalkApplication::class)
+class LeaveConversationWorker(context: Context, workerParams: WorkerParameters) :
+    ListenableWorker(context, workerParams) {
+
+    @Inject
+    lateinit var ncApi: NcApi
+
+    @Inject
+    lateinit var userManager: UserManager
+
+    private val result = SettableFuture.create<Result>()
+
+    override fun startWork(): ListenableFuture<Result> {
+        NextcloudTalkApplication.sharedApplication!!.componentApplication.inject(this)
+        val conversationToken = inputData.getString(BundleKeys.KEY_ROOM_TOKEN)
+        val currentUser = userManager.currentUser.blockingGet()
+
+        if (currentUser != null && conversationToken != null) {
+            val credentials = getCredentials(currentUser.username, currentUser.token)
+            val apiVersion = getConversationApiVersion(currentUser, intArrayOf(ApiUtils.API_V4, 1))
+
+            ncApi.removeSelfFromRoom(
+                credentials,
+                getUrlForParticipantsSelf(apiVersion, currentUser.baseUrl, conversationToken)
+            )
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(object : Observer<GenericOverall?> {
+                    override fun onSubscribe(d: Disposable) {
+                    }
+
+                    override fun onNext(p0: GenericOverall) {
+                    }
+
+                    override fun onError(e: Throwable) {
+                        Log.e(TAG, "Failed to remove self from room", e)
+                        if (e.message?.contains("HTTP 400") == true) {
```

Comment:
For errors other than http400, Result.failure() would not be thrown which could cause some bugs (worker won't finish...).

So we have to make sure to differentiate between the errors. We can do this by passing data to the result, 
so define a consts like:
```
  const val ERROR_NO_OTHER_MODERATORS_OR_OWNERS_LEFT = "NO_OTHER_MODERATORS_OR_OWNERS_LEFT"
  const val ERROR_OTHER = "ERROR_OTHER"
```
and set it as data for error_type (which could also be a const string):

Also instead to analyze the string, let's check the error code

```
  override fun onError(e: Throwable) {
      Log.e(TAG, "Failed to remove self from room", e)
      val httpException = e as? HttpException
      val errorData = if (httpException?.code() == 400) {
          Data.Builder()
              .putString("error_type", ERROR_NO_OTHER_MODERATORS_OR_OWNERS_LEFT)
              .build()
      } else {
          Data.Builder()
              .putString("error_type", ERROR_OTHER)
              .build()
      }
      result.set(Result.failure(errorData))
  }
```

When getting the work info, read the data to identify the error type:

```
WorkInfo.State.FAILED -> {
    val errorType = workInfo.outputData.getString("error_type")

    if (errorType == LeaveConversationWorker.ERROR_NO_OTHER_MODERATORS_OR_OWNERS_LEFT) {
        Snackbar.make(
            binding.root,
            R.string.nc_last_moderator_leaving_room_warning,
            Snackbar.LENGTH_LONG
        ).show()
    } else {
        Snackbar.make(
            binding.root,
            R.string.nc_common_error_sorry,
            Snackbar.LENGTH_LONG
        ).show()
    }
}
```